t/t5799-gvfs-helper.sh

```diff
@@ -0,0 +1,972 @@
+
+# Setup some loopback URLs where test-gvfs-protocol.exe will be
+# listening.  We will spawn it directly inside the repo_src directory,
+# so we don't need any of the directory mapping or configuration
+# machinery found in "git-daemon.exe" or "git-http-backend.exe".
+#
+# This lets us use the "uri-base" part of the URL (prior to the REST
+# API "/gvfs/<token>") to control how our mock server responds.  For
+# example, only the origin (main Git) server supports "/gvfs/config".
+#
+# For example, this means that if we add a remote containing $ORIGIN_URL,
+# it will work with gvfs-helper, but not for fetch (without some mapping
+# tricks).
+#
+HOST_PORT=127.0.0.1:$GIT_TEST_GVFS_PROTOCOL_PORT
+ORIGIN_URL=http://$HOST_PORT/servertype/origin
+CACHE_URL=http://$HOST_PORT/servertype/cache
+
+SHARED_CACHE_T1="$PWD"/shared_cache_t1
+
+# The pid-file is created by test-gvfs-protocol.exe when it starts.
+# The server will shut down if/when we delete it.  (This is a little
+# easier than killing it by PID.)
+#
+PID_FILE="$PWD"/pid-file.pid
+SERVER_LOG="$PWD"/OUT.server.log
+
+PATH="$GIT_BUILD_DIR/t/helper/:$PATH" && export PATH
+
+OIDS_FILE="$PWD"/oid_list.txt
+OIDS_CT_FILE="$PWD"/oid_ct_list.txt
+OIDS_BLOBS_FILE="$PWD"/oids_blobs_file.txt
+OID_ONE_BLOB_FILE="$PWD"/oid_one_blob_file.txt
+
+# Get a list of available OIDs in repo_src so that we can try to fetch
+# them and so that we don't have to hard-code a list of known OIDs.
+# This doesn't need to be a complete list -- just enough to drive some
+# representative tests.
+#
+# Optionally require that we find a minimum number of OIDs.
+#
+get_list_of_oids () {
+	git -C "$REPO_SRC" rev-list --objects HEAD | sed 's/ .*//' | sort >"$OIDS_FILE"
+
+	if test $# -eq 1
+	then
+		actual_nr=$(( $(wc -l <"$OIDS_FILE") ))
+		if test $actual_nr -lt $1
+		then
+			echo "get_list_of_oids: insufficient data.  Need $1 OIDs."
+			return 1
+		fi
+	fi
+	return 0
+}
+
+get_list_of_blobs_oids () {
+	git -C "$REPO_SRC" ls-tree HEAD | grep ' blob ' | awk "{print \$3}" | sort >"$OIDS_BLOBS_FILE"
+	head -1 <"$OIDS_BLOBS_FILE" >"$OID_ONE_BLOB_FILE"
+}
+
+get_list_of_commit_and_tree_oids () {
+	git -C "$REPO_SRC" cat-file --batch-check --batch-all-objects | awk "/commit|tree/ {print \$1}" | sort >"$OIDS_CT_FILE"
+
+	if test $# -eq 1
+	then
+		actual_nr=$(( $(wc -l <"$OIDS_CT_FILE") ))
+		if test $actual_nr -lt $1
+		then
+			echo "get_list_of_commit_and_tree_oids: insufficient data.  Need $1 OIDs."
+			return 1
+		fi
+	fi
+	return 0
+}
+
+test_expect_success 'setup repos' '
+	test_create_repo "$REPO_SRC" &&
+	#
+	# test_commit_bulk() does magic to create a packfile containing
+	# the new commits.
+	#
+	test_commit_bulk -C "$REPO_SRC" --filename="batch_a.%s.t" 9 &&
+	cp "$REPO_SRC"/.git/refs/heads/master m1.branch &&
+	test_commit_bulk -C "$REPO_SRC" --filename="batch_b.%s.t" 9 &&
+	cp "$REPO_SRC"/.git/refs/heads/master m2.branch &&
+	#
+	# test_commit() creates commits, trees, tags, and blobs and leave
+	# them loose.
```

Comment:
You might want to force auto-gc off via `test_config gc.auto 0 &&` so that this promise is definitely kept.

Reply:
But we will be very unlikely to surpass the default limit for triggering a GC (but who knows what happens in the future).

Reply:
True.